### PR TITLE
[clangd] Avoid libFormat's objective-c guessing heuristic where possible

### LINKED ISSUE #719: Extreme memory usage when opening big header files
Clangd uses up all of the available memory if you open big header files until eventually the system freezes
Example header file: [miniaudio.h](https://github.com/mackron/miniaudio/blob/master/miniaudio.h)

**Logs**
clangd output:
```
I[11:50:29.051] ASTWorker building file /.../lib/miniaudio/miniaudio.h version 1 with command inferred from /.../InstanceGuard.cpp
[/.../build]
/usr/bin/c++ --driver-mode=g++ -g -c -x c++-header -std=c++11 /.../lib/miniaudio/miniaudio.h -fsyntax-only -resource-dir=/.../clangd/lib/clang/13.0.0
I[11:50:29.052] --> textDocument/clangd.fileStatus
I[11:50:29.095] --> workspace/semanticTokens/refresh(2)
I[11:50:29.095] <-- reply(2)
I[11:50:29.127] <-- textDocument/codeAction(14)
I[11:50:29.127] <-- $/cancelRequest
I[11:50:29.131] <-- textDocument/semanticTokens/full/delta(15)
I[11:50:29.132] --> reply:textDocument/semanticTokens/full/delta(15) 0 ms
I[11:50:29.132] --> textDocument/clangd.fileStatus
I[11:50:29.846] <-- textDocument/codeAction(16)
I[11:50:29.846] <-- $/cancelRequest
```
There is no error output from the extension or clangd

**System information**
Output of `clangd --version`:
```
> clangd --version
clangd version 13.0.0 (https://github.com/llvm/llvm-project 3a11a41795bec548e91621caaa4cc00fc31b2212)
```
Editor/LSP plugin: [vscode-clangd](https://marketplace.visualstudio.com/items?itemName=llvm-vs-code-extensions.vscode-clangd)
Operating system: Arch Linux (Kernel: 5.11.0-126-tkg-pds)

Related issues:
#251 
#115 
#694

### LINKED ISSUE #1384: clangd unable to parse some header files due to exponential memory usage causing OOM
clangd consumes all memory on my system (~24GiB) before being OOM killed while attempting to parse [this attached header file][hdr] (rename it to `mathlink.h` because GitHub is stupid and won't let me upload a file with a `.h` extension on a web site about programming).

[hdr]: <https://github.com/clangd/clangd/files/10046587/mathlink.txt>

Presumably it's stuck in a loop that would never terminate; or it might if given more memory, but would require an astronomical amount to succeed--notice, however, that the header contents are hardly exotic: just preprocessor junk and C declarations.

Repro with `systemd-run --user -p MemoryMax=8G -S`, then in the shell, run `clangd --log=verbose --check=mathlink.h` (you don't **have** to use `systemd-run`, but beware of systemd/systemd#25376 if you don't!)

**Logs**

[`clangd-stderr.txt`](https://github.com/clangd/clangd/files/10046563/clangd-stderr.txt)

**System information**

Output of `clangd --version`:

```
clangd version 15.0.3 (https://github.com/llvm/llvm-project 4a2c05b05ed07f1f620e94f6524a8b4b2760a0b1)
Features: linux+grpc
Platform: x86_64-unknown-linux-gnu
```

Editor/LSP plugin:

Emacs 28.1 + lsp-mode

Operating system:

```
Linux arch-vm 6.0.7-arch1-1 #1 SMP PREEMPT_DYNAMIC Thu, 03 Nov 2022 18:01:58 +0000 x86_64 GNU/Linux
```

### LINKED ISSUE #70945: clangd uses up all the memory (~90GB) for miniaudio.h
Reproduces on all clangd versions on Linux (tried clangd-14, 16, and current master 6477b41a0b56919c42f44c3499ed32720101b43b).

Steps:
1. Download miniaudio.h: https://github.com/mackron/miniaudio/blob/master/miniaudio.h
2. Open it in VSCode with the clangd extension enabled
3. Watch the memory gets eaten up
Screenshot from 2023-11-01 15-48-30
4. Thank the Linux Kernel for saving the day

```
Nov 01 16:01:09 debian kernel: Out of memory: Killed process 170232 (clangd.main) total-vm:99151264kB, anon-rss:29007340kB, file-rss:3840kB, shmem-rss:0kB, UID:1000 pgtables:189720kB oom_score_adj:200
Nov 01 16:01:09 debian systemd[1]: user@1000.service: A process of this unit has been killed by the OOM killer.
```

This also reproduces for NeoVim with neovim-lspconfig and clang.

### PR DESCRIPTION
**Pull request by HighCommander4**
This avoids a known libFormat bug where the heuristic can OOM on certain large files (particularly single-header libraries such as miniaudio.h).

The OOM will still happen on affected files if you actually try to format them (this is harder to avoid since the underlyting issue affects the actual formatting logic, not just the language-guessing heuristic), but at least it's avoided during non-modifying operations like hover, and modifying operations that do local formatting like code completion.

Fixes https://github.com/clangd/clangd/issues/719
Fixes https://github.com/clangd/clangd/issues/1384
Fixes https://github.com/llvm/llvm-project/issues/70945

### CODE REVIEW

**Comment by HighCommander4** [comment]
For additional context, please refer to the discussion of https://github.com/llvm/llvm-project/pull/78925, which was my first attempt to fix these bugs using a different approach

File: clang-tools-extra/clangd/SourceCode.h [hunk 1]
```diff
@@ -166,14 +166,29 @@ TextEdit toTextEdit(const FixItHint &FixIt, const SourceManager &M,
...
+  // which can be arbitrarily long).
+  EntireFileOrRange
+};
+
 /// Choose the clang-format style we should apply to a certain file.
 /// This will usually use FS to look for .clang-format directories.
 /// FIXME: should we be caching the .clang-format file search?
 /// This uses format::DefaultFormatStyle and format::DefaultFallbackStyle,
 /// though the latter may have been overridden in main()!
 format::FormatStyle getFormatStyleForFile(llvm::StringRef File,
```
Comment (hokein) [inline clang-tools-extra/clangd/SourceCode.h:188]: can we add some unit tests for this function in `SourceCodeTest.cpp`?

**Review by hokein — COMMENTED** [review]
thanks, this looks good in general, just some nits.

File: clang-tools-extra/clangd/SourceCode.h [hunk 2]
```diff
@@ -166,14 +166,29 @@ TextEdit toTextEdit(const FixItHint &FixIt, const SourceManager &M,
...
+enum class FormatKind {
+  // Formatting a snippet of synthesized code (e.g. a code snippet
+  // shown in a hover) that's not part of the main file.
+  Snippet,
+  // Formatting edits made by an editor action such as code completion
+  // or rename.
+  Replacements,
+  // Formatting the entire main file (or a range selected by the user,
+  // which can be arbitrarily long).
+  EntireFileOrRange
```
[the remark this answers is not in the document]
Comment (HighCommander4) [reply clang-tools-extra/clangd/SourceCode.h]: Done, patch uses a bool flag now

File: clang-tools-extra/clangd/SourceCode.h (hunk 1 above)
Comment (HighCommander4) [reply clang-tools-extra/clangd/SourceCode.h:177]: Sure, I added a unit test which checks the detected language in various cases, including checking that the new `FormatFile` flag successfully bypasses the language guessing heuristic

**Pull request merged**